allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -77,21 +77,29 @@ def __init__(self,
         self._beam_search = decoder_beam_search
         self._max_decoding_steps = max_decoding_steps
         self._action_namespace = action_namespace
+        self._action_padding_index = 0 if self.vocab.is_padded(self._action_namespace) else -1
 
         self._start_index = self.vocab.get_token_index(START_SYMBOL, self._action_namespace)
         self._end_index = self.vocab.get_token_index(END_SYMBOL, self._action_namespace)
-        num_actions = self.vocab.get_vocab_size(self._action_namespace)
-        self._decoder_step = WikiTablesDecoderStep(self._encoder.get_output_dim(),
-                                                   action_embedding_dim,
-                                                   num_actions,
-                                                   attention_function,
-                                                   self._start_index)
+        # This is a _global_ mapping because eventually we will have additional allowed actions for
+        # each instance (e.g., instance-specific entities or predicates), and for the current
+        # decoder state (e.g., you can only produce variables inside of a lambda expression).
+        # These are just the actions that are specified by the grammar.
+        self._global_type_productions = _get_type_productions(self.vocab, self._action_namespace)
+        self._decoder_step = WikiTablesDecoderStep(vocab=vocab,
+                                                   action_namespace=action_namespace,
+                                                   encoder_output_dim=self._encoder.get_output_dim(),
+                                                   action_embedding_dim=action_embedding_dim,
+                                                   attention_function=attention_function,
+                                                   start_index=self._start_index)
 
     @overrides
     def forward(self,  # type: ignore
                 question: Dict[str, torch.LongTensor],
+                table: Dict[str, torch.LongTensor],
```

Comment:
Is this for entity linking? Add this to the parameters in docstring.

Reply:
Done.